phoenix-core-client/src/main/java/org/apache/phoenix/schema/CompiledConditionalTTLExpression.java

```diff
@@ -0,0 +1,307 @@
+    // first re-written to use index column references and then compiled.
+    private final Expression compiledExpr;
+    // columns referenced in the ttl expression to be added to scan
+    private final Set<ColumnReference> conditionExprColumns;
+
+    // Since DeleteColumn updates are not propagated to indexes we mask older columns
+    // within GlobalIndexChecker. We need to do the same masking in TTLRegionScanner
+    // before we evaluate the ttl expression on the row. We apply this masking on server
+    // maintained indexes.
+    private final boolean maskOlderCells;
+
+    public CompiledConditionalTTLExpression(String ttlExpr,
+                                            Expression compiledExpression,
+                                            Set<ColumnReference> conditionExprColumns,
+                                            boolean maskOlderCells) {
+        Preconditions.checkNotNull(compiledExpression);
+        Preconditions.checkNotNull(conditionExprColumns);
+        this.ttlExpr = ttlExpr;
+        this.compiledExpr = compiledExpression;
+        this.conditionExprColumns = conditionExprColumns;
+        this.maskOlderCells = maskOlderCells;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        CompiledConditionalTTLExpression that = (CompiledConditionalTTLExpression) o;
+        return ttlExpr.equals(that.ttlExpr);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(ttlExpr);
+    }
+
+    @Override
+    public String getTTLExpression() {
+        return ttlExpr;
+    }
+
+    @Override
+    public String toString() {
+        return getTTLExpression();
+    }
+
+    /**
+     * The cells of the row (i.e., result) read from HBase store are lexicographically ordered
+     * for tables using the key part of the cells which includes row, family, qualifier,
+     * timestamp and type. The cells belong of a column are ordered from the latest to
+     * the oldest. The method leverages this ordering and groups the cells into their columns
+     * based on the pair of family name and column qualifier.
+     */
+    private List<Cell> getLatestRowVersion(List<Cell> result) {
+        List<Cell> latestRowVersion = new ArrayList<>();
+        Cell currentColumnCell = null;
+        long maxDeleteFamilyTS = 0;
+        for (Cell cell : result) {
+            if (currentColumnCell == null ||
+                    !CellUtil.matchingColumn(cell, currentColumnCell)) {
+                // found a new column cell which has the latest timestamp
+                currentColumnCell = cell;
+                if (currentColumnCell.getType() == Cell.Type.DeleteFamily ||
+                        currentColumnCell.getType() == Cell.Type.DeleteFamilyVersion) {
```

Comment:
DeleteFamilyVersion should not be treated as DeleteFamily here. In other words, DeleteFamilyVersion deletes only a specific version of the row such that the timestamp of the DeleteFamilyVersion equals to the timestamp of the row version.